WindowManager/app/src/main/java/com/example/windowmanagersample/backend/MidScreenFoldBackend.kt

```diff
@@ -18,105 +18,139 @@ package com.example.windowmanagersample.backend
- * <p> If the application window is portrait-oriented, then the reported fold in the screen will
- * be horizontal (vertically-folding device configuration). If the window is landscape-oriented,
- * then the reported fold will be vertical (horizontally-folding device configuration).
- * @see [WindowManager]
- * @see [getWindowLayoutInfo]
+ * A sample backend that will have a fold in the middle of the screen. The {@link FoldAxis}
+ * specifies which axis is followed. This sample backend can be used to model devices that open
+ * like a clam shell or a book. This is relative to the display's dimensions. For displays that
+ * are taller than they are wide with a short fold axis they will mimic a horizontal fold. For
+ * displays that are wider than they are tall with a short fold axis will mimic a vertical fold.
+ *
+ * <p>The {@link DeviceState} is fixed to have an opened posture. This sample implementation
+ * ignores the device state listener. Combine with {@link InitialValueWindowBackendDecorator} to
+ * receive the initial value through the listener.
+ *
+ * <p>The {@link WindowLayoutInfo} is also fixed to have a {@link TYPE_FOLD} where the longest
+ * dimension runs parallel to the specified {@link FoldAxis}. The fold is placed in the middle
+ * of the display. Like the {@link DeviceState}, the sample implementation ignores the window
+ * layout info listener. Combine with {@link InitialValueWindowBackendDecorator} to
+ * receive the initial value through the listener.
  */
-class MidScreenFoldBackend : WindowBackend {
-
-    private var deviceState = DeviceState.Builder().setPosture(DeviceState.POSTURE_OPENED).build()
+class MidScreenFoldBackend(private val foldAxis: FoldAxis) : WindowBackend {
+    /**
+     * The side which the fold axis should be parallel to.
+     */
+    enum class FoldAxis {
+        LONG_DIMENSION,
+        SHORT_DIMENSION
+    }
 
+    private var deviceState = FoldingFeature.STATE_FLAT
     private var windowLayoutInfoCallback: Consumer<WindowLayoutInfo>? = null
-    private var deviceStateCallback: Consumer<DeviceState>? = null
-
-    override fun getDeviceState() = deviceState
+    private var windowLayoutInfoExecutor: Executor? = null
 
     override fun getWindowLayoutInfo(context: Context): WindowLayoutInfo {
-        val activity = getActivityFromContext(context) ?: throw IllegalArgumentException(
-            "Used non-visual Context used with WindowManager. Please use an Activity or a " +
-                "ContextWrapper around an Activity instead."
-        )
-        val displaySize = Point(activity.window.decorView.width, activity.window.decorView.height)
-        val featureRect = if (displaySize.x >= displaySize.y) { // Landscape
-            Rect(displaySize.x / 2, 0, displaySize.x / 2, displaySize.y)
-        } else { // Portrait
-            Rect(0, displaySize.y / 2, displaySize.x, displaySize.y / 2)
-        }
+        val windowSize = (context as Activity).calculateWindowSizeExt()
```

Comment:
Commented on this in the activity extensions file, we should be able to use currentWindowMetrics here instead

Reply:
+1